### ISSUE TITLE
SNOW-894842: Show the underlying error message when reading a config file fails

### ISSUE THREAD

**Issue by sfc-gh-zblackwood** [snowflakedb/snowflake-connector-python#1696]
### What is the current behavior?

When using a tool like snowcli, if there is an error in ~/.snowflake/config.toml, the actual underlying error message is hidden, leaving only `"An unknown error happened while loading <filename>"`, leaving the user unable to determine whether this is an error with the toml file existing, an access error, or some error with the contents itself.

### What is the desired behavior?

Append the actual error message contents to the ConfigSourceError.

### How would this improve `snowflake-connector-python`?

It would make it easier to diagnose the sources of configuration issues, and determine what sorts of errors they are. While, if you are using snowflake-connector-python directly you might see the actual error message higher in the stack trace, if using a tool like snowcli the actual underly error is completely hidden. For example, this is all that I see if I add a syntax error in the toml.

```
snow --debug connection test
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/zblackwood/.local/pipx/venvs/snowflake-cli-labs/lib/python3.11/site-packages/snowcli/cli/ │
│ app.py:119 in default                                                                            │
│                                                                                                  │
│   116 │   │   pycharm_debug_server_host=pycharm_debug_server_host,                               │
│   117 │   │   pycharm_debug_server_port=pycharm_debug_server_port,                               │
│   118 │   )                                                                                      │
│ ❱ 119 │   config_init(configuration_file)                                                        │
│   120 │   loggers.create_loggers(verbose, debug)                                                 │
│   121 │   setup_global_context(debug=debug)                                                      │
│   122                                                                                            │
│                                                                                                  │
│ /Users/zblackwood/.local/pipx/venvs/snowflake-cli-labs/lib/python3.11/site-packages/snowcli/conf │
│ ig.py:139 in config_init                                                                         │
│                                                                                                  │
│   136 │   Initializes the app configuration. Config provided via cli flag takes precedence.      │
│   137 │   If config file does not exist we create an empty one.                                  │
│   138 │   """                                                                                    │
│ ❱ 139 │   cli_config.from_context(config_path_override=config_file)                              │
│   140                                                                                            │
│   141                                                                                            │
│   142 cli_config: CliConfigManager = CliConfigManager()  # type: ignore                          │
│                                                                                                  │
│ /Users/zblackwood/.local/pipx/venvs/snowflake-cli-labs/lib/python3.11/site-packages/snowcli/conf │
│ ig.py:31 in from_context                                                                         │
│                                                                                                  │
│    28 │   │   │   self.file_path = config_path_override                                          │
│    29 │   │   if not self.file_path.exists():                                                    │
│    30 │   │   │   self._initialise_config()                                                      │
│ ❱  31 │   │   self.read_config()                                                                 │
│    32 │                                                                                          │
│    33 │   def get_section(self, *path) -> dict:                                                  │
│    34 │   │   section = self._find_section(*path)                                                │
│                                                                                                  │
│ /Users/zblackwood/.local/pipx/venvs/snowflake-cli-labs/lib/python3.11/site-packages/snowflake/co │
│ nnector/config_manager.py:285 in read_config                                                     │
│                                                                                                  │
│   282 │   │   │   try:                                                                           │
│   283 │   │   │   │   read_config_piece = tomlkit.parse(filep.read_text())                       │
│   284 │   │   │   except Exception as e:                                                         │
│ ❱ 285 │   │   │   │   raise ConfigSourceError(                                                   │
│   286 │   │   │   │   │   "An unknown error happened while loading " f"'{str(filep)}'"           │
│   287 │   │   │   │   ) from e                                                                   │
│   288 │   │   │   if section is None:                                                            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ConfigSourceError: An unknown error happened while loading '/Users/zblackwood/.snowflake/config.toml'
```

If the text of the ConfigSourceError included the original error, I would get this error:

```
ConfigSourceError: An unknown error happened while loading '/Users/zblackwood/.snowflake/config.toml': Control characters (codes less than 0x1f and 0x7f) are not allowed in strings, use \u000a instead at line 3 col 18
```

Still not the most clear error, but at least it tells me that there is some issue with the contents of the file, and it is on line 3, col 18.

**Comment by sfc-gh-mkeller** [comment]
@sfc-gh-zblackwood Look at the code in your error-message: https://github.com/snowflakedb/snowflake-connector-python/blob/20f1efb54ba07f2ee2e632dbc017c7d3807a2554/src/snowflake/connector/config_manager.py#L284C12-L284C12

We utilize [exception chaining](https://docs.python.org/3/tutorial/errors.html#exception-chaining) instead of appending exception messages. 

@sfc-gh-turbaszek could you see if there was a better way to make `typer` work with exception chaining?

**Comment by sfc-gh-zblackwood** [comment]
@sfc-gh-mkeller Yes, I saw that, but noticed that the underlying exception wasn't shown, and figured this was a good way to get that working. I tried a simple example with Typer, and it does indeed show the original exception -- apparently showing chained exceptions is part of the default behavior of Typer + Rich. I tested with this example 
```python
import typer


def main(name: str = "morty"):
    try:
        print(name + 3)
    except Exception as e:
        raise ValueError("An error") from e


if __name__ == "__main__":
    typer.run(main)
```

And the original exception showed properly. So, it seems like this is more of a configuration issue with snowcli. I'll move this issue over to there.
